ticdc: refine the ticdc changefeed configuration file (#14014)
pingcap/docs-cn: ticdc/ticdc-changefeed-config.md

本章节详细介绍了同步任务的配置。

```toml

> # 指定该 Changefeed 在 Capture Server 中的内存配额的上限。对于超额使用部分，
> # 会在运行中被 go runtime 优先回收，默认值为 `1073741824` (即 1 GB)。

Reviewer: ```suggestion
# 指定该 Changefeed 在 Capture Server 中内存配额的上限。对于超额使用部分，
# 会在运行中被 Go runtime 优先回收。默认值为 `1073741824`，即 1 GB。
```

> # 指定该 changefeed 在 capture server 中的内存配额的上限。对于超额使用部分，
> # 会在运行中被 go runtime 优先回收，默认为 1GB
> # memory-quota = 1073741824 # 1 GB 

Reviewer 2: ```suggestion
# 指定该 changefeed 在 capture server 中的内存配额的上限。对于超额使用部分，
# 会在运行中被 go runtime 优先回收，默认值为 `1073741824` (即 1 GB)。
# memory-quota = 1073741824
```

enable-old-value = true

# 是否开启 Syncpoint 功能，从 v6.3.0 开始支持，该功能默认关闭。
# 从 v6.4.0 开始，使用 Syncpoint 功能需要同步任务拥有下游集群的 SYSTEM_VARIABLES_ADMIN 或者 SUPER 权限。

> # 注意：该参数只有当下游为 Kafka 或 Storage Service 时，才会生效。

Reviewer 2: ```suggestion
# 注意：该参数只有当下游为 Kafka 或存储服务时，才会生效。
```

# enable-sync-point = false

# Syncpoint 功能对齐上下游 snapshot 的时间间隔
# 配置格式为 h m s，例如 "1h30m30s"
# 默认值为 10m，最小值为 30s

> # 注意：该参数只有当下游为 Kafka 或 Storage Service 时，才会生效。

Reviewer 2: ```suggestion
# 注意：该参数只有当下游为 Kafka 或存储服务时，才会生效。
```

# sync-point-interval = "5m"

# Syncpoint 功能在下游表中保存的数据的时长，超过这个时间的数据会被清理
# 配置格式为 h m s，例如 "24h30m30s"
# 默认值为 24h

> # 注意：该参数只有当下游为 Kafka 或 Storage Service 时，才会生效。

Reviewer 2: ```suggestion
# 注意：该参数只有当下游为 Kafka 或存储服务时，才会生效。
```

# 过滤规则语法：https://docs.pingcap.com/zh/tidb/stable/table-filter#表库过滤语法
rules = ['*.*', '!test.*']

# 忽略特定 start_ts 的事务
# 默认值为空列表。

> # IgnoreTxnStartTs = [1683225923]

Reviewer: ```suggestion
# IgnoreTxnStartTs = []
```
示例中使用默认值

# 对于 MQ 类的 Sink，可以通过 dispatchers 配置 event 分发器
# 支持 partition 及 topic（从 v6.1 开始支持）两种 event 分发器。二者的详细说明见下一节。
# matcher 的匹配语法和过滤器规则语法相同，matcher 匹配规则的详细说明见下一节。

> # 注意：该参数只有当下游为 MQ 时，才会生效。

Reviewer 2: ```suggestion
# 注意：该参数只有当下游为消息队列时，才会生效。
```

# protocol 用于指定传递到下游的协议格式
# 当下游类型是 Kafka 时，支持 canal-json、avro 两种协议。
# 当下游类型是存储服务时，目前仅支持 canal-json、csv 两种协议。
# 注意：该字段只有当下游为 Kafka 或 Storage Service 时，才会生效。
# protocol = "canal-json"

# 以下三个配置项仅在同步到存储服务的 sink 中使用，在 MQ 和 MySQL 类 sink 中无需设置。
# 换行符，用来分隔两个数据变更事件。默认值为空，表示使用 "\r\n" 作为换行符。
# terminator = ''
# 文件路径的日期分隔类型。可选类型有 `none`、`year`、`month` 和 `day`。默认值为 `none`，即不使用日期分隔。详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。
# 注意：该字段只有当下游为 Storage Service 时，才会生效。
date-separator = 'none'
# 是否使用 partition 作为分隔字符串。默认值为 true，即一张表中各个 partition 的数据会分不同的目录来存储。建议保持该配置项为 true 以避免下游分区表可能丢数据的问题 <https://github.com/pingcap/tiflow/issues/8581>。使用示例详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。
# 注意：该字段只有当下游为 Storage Service 时，才会生效。
enable-partition-separator = true

> 

Reviewer 2: 下面部分漏掉了
```suggestion

# Schema registry URL. Note: ignore this field if you replicate data to the downstream types other than MQ. 
# schema-registry = "http://localhost:80801/subjects/{subject-name}/versions/{version-number}/schema"

# Specifies the number of encoder threads used when encoding data.
# Note: ignore this field when the downstream type is not the MQ.
# The default value is 16.
# encoder-concurrency = 16

# Specifies whether to enable kafka-sink-v2 that uses the kafka-go sink library.
# Note: ignore this field when the downstream type is not the MQ.
# The default value is false.
# enable-kafka-sink-v2 = false

# Specifies whether to only output the updated columns.
# Note: this field only applies to the MQ downstream using the open-protocol and canal-json.
# The default value is false.
# only-output-updated-columns = false

```

Author: added

# protocol 用于指定传递到下游的协议格式
# 当下游类型是 Kafka 时，支持 canal-json、avro 两种协议。
# 当下游类型是存储服务时，目前仅支持 canal-json、csv 两种协议。
# 注意：该字段只有当下游为 Kafka 或 Storage Service 时，才会生效。
# protocol = "canal-json"

# 以下三个配置项仅在同步到存储服务的 sink 中使用，在 MQ 和 MySQL 类 sink 中无需设置。
# 换行符，用来分隔两个数据变更事件。默认值为空，表示使用 "\r\n" 作为换行符。
# terminator = ''
# 文件路径的日期分隔类型。可选类型有 `none`、`year`、`month` 和 `day`。默认值为 `none`，即不使用日期分隔。详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。
# 注意：该字段只有当下游为 Storage Service 时，才会生效。
date-separator = 'none'
# 是否使用 partition 作为分隔字符串。默认值为 true，即一张表中各个 partition 的数据会分不同的目录来存储。建议保持该配置项为 true 以避免下游分区表可能丢数据的问题 <https://github.com/pingcap/tiflow/issues/8581>。使用示例详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。
# 注意：该字段只有当下游为 Storage Service 时，才会生效。
enable-partition-separator = true

# Schema registry URL。注意： 该字段只有当下游为消息队列时，才会生效。
# schema-registry = "http://localhost:80801/subjects/{subject-name}/versions/{version-number}/schema"

> # Specifies the number of encoder threads used when encoding data.
> # encoder 线程数。注意： 该字段只有当下游为消息队列时，才会生效。默认值为16

Reviewer 2: ```suggestion
# 指定编码数据时使用的编码器的线程数。
# 默认值为16。
# 注意： 该字段只有当下游为消息队列时，才会生效。
```

# encoder-concurrency = 16

> # Specifies whether to enable kafka-sink-v2 that uses the kafka-go sink library.
> # 声明是否开启 kafka-sink-v2，kafka-sink-v2 内部使用 kafka-go 实现。

Reviewer 2: ```suggestion
# Specifies whether to enable kafka-sink-v2 that uses the kafka-go sink library.
# 是否开启 kafka-sink-v2，kafka-sink-v2 内部使用 kafka-go 实现。
```

> # 注意： 该字段只有当下游为消息队列时，才会生效。默认值为 false。
> # enable-kafka-sink-v2 = false
> # 声明是否只向下游同步有内容更新的列
> # 注意： 该字段只有当下游为消息队列，并且使用 open-protocol 或者 canal-json 时，才会生效。默认值为 false。

Reviewer 2: ```suggestion
# 注意：该字段只有当下游为消息队列时，才会生效。默认值为 false。
# enable-kafka-sink-v2 = false
# 是否只向下游同步有内容更新的列。
# 注意：该字段只有当下游为消息队列，并且使用 open-protocol 或者 canal-json 时，才会生效。默认值为 false。
```

# protocol 用于指定传递到下游的协议格式
# 当下游类型是 Kafka 时，支持 canal-json、avro 两种协议。
# 当下游类型是存储服务时，目前仅支持 canal-json、csv 两种协议。
# 注意：该字段只有当下游为 Kafka 或 Storage Service 时，才会生效。
# protocol = "canal-json"

# 以下三个配置项仅在同步到存储服务的 sink 中使用，在 MQ 和 MySQL 类 sink 中无需设置。
# 换行符，用来分隔两个数据变更事件。默认值为空，表示使用 "\r\n" 作为换行符。
# terminator = ''
# 文件路径的日期分隔类型。可选类型有 `none`、`year`、`month` 和 `day`。默认值为 `none`，即不使用日期分隔。详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。
# 注意：该字段只有当下游为 Storage Service 时，才会生效。
date-separator = 'none'
# 是否使用 partition 作为分隔字符串。默认值为 true，即一张表中各个 partition 的数据会分不同的目录来存储。建议保持该配置项为 true 以避免下游分区表可能丢数据的问题 <https://github.com/pingcap/tiflow/issues/8581>。使用示例详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。
# 注意：该字段只有当下游为 Storage Service 时，才会生效。
enable-partition-separator = true

# Schema registry URL。注意： 该字段只有当下游为消息队列时，才会生效。
# schema-registry = "http://localhost:80801/subjects/{subject-name}/versions/{version-number}/schema"
# 编码数据时使用的编码器的线程数。
# 默认值为 16。
# 注意： 该字段只有当下游为消息队列时，才会生效。
# encoder-concurrency = 16

> # Specifies whether to enable kafka-sink-v2 that uses the kafka-go sink library.

Reviewer 2: ```suggestion
```

# protocol 用于指定传递到下游的协议格式
# 当下游类型是 Kafka 时，支持 canal-json、avro 两种协议。
# 当下游类型是存储服务时，目前仅支持 canal-json、csv 两种协议。

> # 注意：该参数只有当下游为 Kafka 或 Storage Service 时，才会生效。

Reviewer: ```suggestion
# 注意：该参数只有当下游为 Kafka 或存储服务时，才会生效。
```

# protocol = "canal-json"

# 以下三个配置项仅在同步到存储服务的 sink 中使用，在 MQ 和 MySQL 类 sink 中无需设置。
# 换行符，用来分隔两个数据变更事件。默认值为空，表示使用 "\r\n" 作为换行符。
# terminator = ''

# 文件路径的日期分隔类型。可选类型有 `none`、`year`、`month` 和 `day`。默认值为 `none`，即不使用日期分隔。详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。

> # 注意：该参数只有当下游为 Storage Service 时，才会生效。

Reviewer: ```suggestion
# 注意：该参数只有当下游为存储服务时，才会生效。
```

date-separator = 'none'

# 是否使用 partition 作为分隔字符串。默认值为 true，即一张表中各个 partition 的数据会分不同的目录来存储。建议保持该配置项为 true 以避免下游分区表可能丢数据的问题 <https://github.com/pingcap/tiflow/issues/8581>。使用示例详见 <https://docs.pingcap.com/zh/tidb/dev/ticdc-sink-to-cloud-storage#数据变更记录>。

> # 注意：该参数只有当下游为 Storage Service 时，才会生效。

Reviewer: ```suggestion
# 注意：该参数只有当下游为存储服务时，才会生效。
```

enable-partition-separator = true

> # Schema registry URL。
> # 注意：该参数只有当下游为消息队列时，才会生效。
> # schema-registry = "http://localhost:80801/subjects/{subject-name}/versions/{version-number}/schema"
> 
> # 编码数据时使用的编码器的线程数。
> # 默认值为 16。
> # 注意：该参数只有当下游为消息队列时，才会生效。
> # encoder-concurrency = 16
> 
> # 是否开启 kafka-sink-v2，kafka-sink-v2 内部使用 kafka-go 实现。
> # 注意：该参数只有当下游为消息队列时，才会生效。
> # 默认值为 false。
> # enable-kafka-sink-v2 = false
> 
> # 是否只向下游同步有内容更新的列。
> # 注意：该参数只有当下游为消息队列，并且使用 open-protocol 或者 canal-json 时，才会生效。
> # 默认值为 false。
> # only-output-updated-columns = false

Reviewer: ```suggestion
# Schema 注册表的 URL。
# 注意：该参数只有当下游为消息队列时，才会生效。
# schema-registry = "http://localhost:80801/subjects/{subject-name}/versions/{version-number}/schema"

# 编码数据时所用编码器的线程数。
# 默认值为 16。
# 注意：该参数只有当下游为消息队列时，才会生效。
# encoder-concurrency = 16

# 是否开启 Kafka Sink V2。Kafka Sink V2 内部使用 kafka-go 实现。
# 默认值为 false。
# 注意：该参数只有当下游为消息队列时，才会生效。
# enable-kafka-sink-v2 = false

# 是否只向下游同步有内容更新的列。
# 默认值为 false。
# 注意：该参数只有当下游为消息队列，并且使用 Open Protocol 或 Canal-JSON 时，才会生效。
# only-output-updated-columns = false
```

# null = '\N'
# 是否在 CSV 行中包含 commit-ts。默认值为 false。
# include-commit-ts = false

> # sink.consistent 包含字段可用来配置 changefeed 的数据一致性。详细的信息，请参考 https://docs.pingcap.com/tidb/stable/ticdc-sink-to-mysql#eventually-consistent-replication-in-disaster-scenarios. 注意：一致性相关字段只有当下游为数据库并且开启 redo log 功能时，才会生效。

Reviewer 2: ```suggestion
# sink.consistent 包含字段可用来配置 changefeed 的数据一致性。详细的信息，请参考 https://docs.pingcap.com/tidb/stable/ticdc-sink-to-mysql#eventually-consistent-replication-in-disaster-scenarios. 
# 注意：一致性相关字段只有当下游为数据库并且开启 redo log 功能时，才会生效。
```

[sink.consistent]

> # 数据一致性级别。可选级别为 none 和 eventual。设置为 none 时，redo log 将被关闭。默认值为 none。
> level = "none"
> # redo log 的最大日志大小，单位为 megabytes。默认值为 64。
> max-log-size = 64
> # 两次 redo log 的刷新间隔，单位为毫秒。默认值为2000。
> flush-interval = 2000 
> # redo log 使用存储服务的 URI。默认值为空。
> storage = "" 
> # 是否将 redo log 存到文件中。默认值为 false。
> use-file-backend = false

Reviewer 2: ```suggestion
# 数据一致性级别。可选级别为 "none" 和 "eventual"。设置为 "none" 时，redo log 将被关闭。默认值为 "none"。
level = "none"
# redo log 的最大日志大小，单位为 megabytes。默认值为 64。
max-log-size = 64
# 两次 redo log 的刷新间隔，单位为毫秒。默认值为 2000。
flush-interval = 2000 
# redo log 使用存储服务的 URI。默认值为空。
storage = "" 
# 是否将 redo log 存到文件中。默认值为 false。
use-file-backend = false
```

# sink.integrity 所包含字段可用于配置数据完整性检验。注意：该字段只有当下游为消息队列时，才会生效。
[sink.integrity]

> # 数据完整性级别。可选项为 none 和 correctness。当设为 none，将关闭数据完整性校验。当设为 correctness，将会对每一行数据进行完整性校验。默认值为 none。
> integrity-check-level = "none"
> # 当检测到数据不完整后的系统行为。可选项包括 warn 和 error。当设为 warn，会将不完整事件记录到日志中并发送到下游。当设为 error，会停止同步数据。默认值为 warn。
> corruption-handle-level = "warn"

Reviewer 2: ```suggestion
# 数据完整性级别。可选项为 none 和 correctness。当设为 none，将关闭数据完整性校验。当设为 correctness，将会对每一行数据进行完整性校验。默认值为 none。
integrity-check-level = "none"
# 当检测到数据不完整后的系统行为。可选项包括 "warn" 和 "error"。当设为 "warn"，会将不完整事件记录到日志中并发送到下游。当设为 "error"，会停止同步数据。默认值为 "warn"。
corruption-handle-level = "warn"
```

# null = '\N'
# 是否在 CSV 行中包含 commit-ts。默认值为 false。
# include-commit-ts = false

> # sink.consistent 中的字段用于配置 Changefeed 的数据一致性。详细的信息，请参考 <https://docs.pingcap.com/tidb/stable/ticdc-sink-to-mysql#eventually-consistent-replication-in-disaster-scenarios>

Reviewer: ```suggestion
# sink.consistent 中的字段用于配置 Changefeed 的数据一致性。详细的信息，请参考 <https://docs.pingcap.com/tidb/stable/ticdc-sink-to-mysql#eventually-consistent-replication-in-disaster-scenarios>。
```

# null = '\N'
# 是否在 CSV 行中包含 commit-ts。默认值为 false。
# include-commit-ts = false

> # sink.consistent 包含字段可用来配置 Changefeed 的数据一致性。详细的信息，请参考 https://docs.pingcap.com/tidb/stable/ticdc-sink-to-mysql#eventually-consistent-replication-in-disaster-scenarios.
> # 注意：一致性相关参数只有当下游为数据库并且开启 redo log 功能时，才会生效。
> [sink.consistent]
> # 数据一致性级别。可选级别为 "none" 和 "eventual"。设置为 "none" 时，redo log 将被关闭。默认值为 "none"。
> level = "none"
> # redo log 的最大日志大小，单位为 MB。默认值为 64。
> max-log-size = 64
> # 两次 redo log 的刷新间隔，单位为毫秒。默认值为 2000。
> flush-interval = 2000
> # redo log 使用存储服务的 URI。默认值为空。
> storage = ""
> # 是否将 redo log 存到文件中。默认值为 false。
> use-file-backend = false

Reviewer: ```suggestion
# sink.consistent 中的字段用于配置 Changefeed 的数据一致性。详细的信息，请参考 <https://docs.pingcap.com/tidb/stable/ticdc-sink-to-mysql#eventually-consistent-replication-in-disaster-scenarios>
# 注意：一致性相关参数只有当下游为数据库并且开启 redo log 功能时，才会生效。
[sink.consistent]
# 数据一致性级别。默认值为 "none"，可选值为 "none" 和 "eventual"。
# 设置为 "none" 时将关闭 redo log。
level = "none"
# redo log 的最大日志大小，单位为 MB。默认值为 64。
max-log-size = 64
# 两次 redo log 刷新的时间间隔，单位为毫秒。默认值为 2000。
flush-interval = 2000
# redo log 使用存储服务的 URI。默认值为空。
storage = ""
# 是否将 redo log 存储到文件中。默认值为 false。
use-file-backend = false
```

> 
> # sink.integrity 所包含参数可用于配置数据完整性检验。
> # 注意：完整性检查相关参数只有当下游为消息队列时，才会生效。
> [sink.integrity]
> # 数据完整性级别。可选项为 "none" 和 "correctness"。
> # 当设为 "none"，将关闭数据完整性校验。
> # 当设为 "correctness"，将会对每一行数据进行完整性校验。
> # 默认值为 "none"。
> integrity-check-level = "none"
> 
> # 当检测到数据不完整后的系统行为。可选项包括 "warn" 和 "error"。
> # 当设为 "warn"，会将不完整事件记录到日志中并发送到下游。
> # 当设为 "error"，会停止同步数据。
> # 默认值为 "warn"。
> corruption-handle-level = "warn"

Reviewer: ```suggestion
```
sink.integrity 是 7.1 引入的，已经添加到 https://github.com/pingcap/docs-cn/pull/13664 中
